// react-hash-routing/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Route = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; } /**
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                * Created by AntonioGiordano on 06/07/16.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                */

var HashRouter = function (_React$Component) {
  _inherits(HashRouter, _React$Component);

  function HashRouter(props) {
    _classCallCheck(this, HashRouter);

    var _this = _possibleConstructorReturn(this, (HashRouter.__proto__ || Object.getPrototypeOf(HashRouter)).call(this, props));

    _this._childKey = null;
    _this._params = [];
    _this.state = {
      style: {
        opacity: 0
      }
    };
    return _this;
  }

  _createClass(HashRouter, [{
    key: 'getDefaultProps',
    value: function getDefaultProps() {
      return {
        onLocationChanged: function onLocationChanged(childKey, params, cb) {
          return cb();
        },
        className: ''
      };
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      window.addEventListener("hashchange", this.onHashChange.bind(this), false);
      this.onHashChange();
    }
  }, {
    key: 'onHashChange',
    value: function onHashChange() {
      var _this2 = this;

      var comp = this;
      this.setState({
        style: {
          opacity: 0
        }
      });
      var ret = this._matchedPage();
      if (ret !== null) {
        this.props.onLocationChanged(this._childKey, this._params, function () {
          setTimeout(function () {
            _this2.setState({
              style: {
                transition: 'opacity 500ms',
                opacity: 1
              }
            });
          }, 100);
          comp.forceUpdate();
        });
      }
    }
  }, {
    key: '_matchedPage',
    value: function _matchedPage() {
      var hash = typeof window.location !== 'undefined' ? window.location.hash : '#/';
      var locArray = hash.split('/');
      if (locArray.length > 0) {
        if (locArray[0] === '#') {
          locArray.shift();
        }
      }
      var locations = [];
      var scores = [];
      var childes = [];
      var params = [];
      _react2.default.Children.map(this.props.children, function (child) {
        var childArray = child.props['hash'].split('/');
        if (childArray.length > 0) {
          childArray.shift();
        }
        locations.push(childArray);
        scores.push(0);
        childes.push(child);
        params.push({});
      });

      var i;
      for (i = 0; i < locations.length; i++) {
        if (locations[i].length !== locArray.length) {
          locations.splice(i, 1);
          scores.splice(i, 1);
          childes.splice(i, 1);
          params.splice(i, 1);
          i--;
        }
      }
      var regexParam = /^{(.*)}$/;
      for (i = 0; i < locArray.length; i++) {
        for (var j = 0; j < locations.length; j++) {
          if (locArray[i] === locations[j][i]) {
            scores[j] += 100;
          } else if (locations[j][i].match(regexParam, '$1') !== null) {
            scores[j] += 1;
            params[j][locations[j][i].match(regexParam, '$1')[1]] = locArray[i];
          } else {
            locations.splice(j, 1);
            scores.splice(j, 1);
            childes.splice(j, 1);
            params.splice(j, 1);
            j--;
          }
        }
      }

      if (locations.length > 0) {
        var max = 0;
        var maxId = 0;
        for (i = 0; i < scores.length; i++) {
          if (scores[i] > max) {
            max = scores[i];
            maxId = i;
          }
        }

        this._childKey = childes[maxId].key;
        this._params = params[maxId];

        return childes[maxId];
      } else {
        return null;
      }
    }
  }, {
    key: 'componentWillUpdate',
    value: function componentWillUpdate() {
      return false;
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.createElement(
        'div',
        { className: this.props.className, style: this.state.style },
        this._matchedPage()
      );
    }
  }]);

  return HashRouter;
}(_react2.default.Component);

HashRouter.propTypes = {
  onLocationChanged: _propTypes2.default.func,
  className: _propTypes2.default.any
};

var Route = exports.Route = function (_React$Component2) {
  _inherits(Route, _React$Component2);

  function Route(props) {
    _classCallCheck(this, Route);

    return _possibleConstructorReturn(this, (Route.__proto__ || Object.getPrototypeOf(Route)).call(this, props));
  }

  _createClass(Route, [{
    key: 'render',
    value: function render() {
      return this.props.children;
    }
  }]);

  return Route;
}(_react2.default.Component);

Route.propTypes = {
  key: _propTypes2.default.string.isRequired,
  hash: _propTypes2.default.string.isRequired
};

exports.default = HashRouter;
